### ISSUE TITLE
New snakeyaml vulnerability has dropped, affects jruby-complete

### ISSUE THREAD

**Issue by apurtell** [jruby/jruby#7570]
jruby-complete bundles snakeyaml, which is receiving considerable attention at the moment because it contains an as of yet unpatched java serialization related arbitrary code execution vulnerability. Please see https://nvd.nist.gov/vuln/detail/CVE-2022-1471 and https://security.snyk.io/vuln/SNYK-JAVA-ORGYAML-3152153 for more information. 

According to Snyk the snakeyaml project has committed a fix to their master branch but has not released yet. Until a fix version is released jruby-complete artifacts are transitively vulnerable to the issue and will be treated by security scanners as such. 

Please update and release jruby-complete when a fixed version is available.

**Comment by headius** [comment]
Thanks for the issue! We'll get Psych fixed and updated and new releases of JRuby out soon.

**Comment by headius** [comment]
Ok I looked into this again.

JRuby is not affected by any of the object deserialization exploits reported against SnakeYAML because we do not use SnakeYAML object serialization support. JRuby utilizes the parser from SnakeYAML directly, passing its events back to Ruby code in Psych. Psych has already made several changes and releases to limit or disable object deserialization by default.

There's really nothing for us to fix here.

@asomov There's no chance of instantiating a Java object if we use ParserImpl directly, is there? Also... my condolences about these false positives... they affect us too, as you see here.

**Comment by apurtell** [comment]
It's not a question if Jruby is vulnerable itself, unfortunately. The presence of vulnerable snakeyaml bytecode bundled in jruby-complete is enough to fail compliance with software bill of material requirements. The 'fix' as such is to rebundle jruby-complete when snakeyaml produces a release containing a fix for the real issue.

**Comment by apurtell** [comment]
For what it's worth, this is not the first instance of bundling in jruby-complete being a problem transitively. BouncyCastle had vulnerabilities reported and jruby-complete bundled that too.

**Comment by headius** [comment]
> It's not a question if Jruby is vulnerable itself, unfortunately.

Unfortunately, there's still nothing for us to do here.

* The Psych library does not use the affected code
* No SnakeYAML release disables the affected code.

If at some point SnakeYAML does release a version that disables or removes the affected code, and we upgrade to it, there will be zero changes in Psych since it does not use that code.

In addition, though we ship SnakeYAML as part of the Psych gem, the jar is only loaded into our internal classloader and cannot really be used from the jruby-complete jar unless a user does the same. SnakeYAML is literally a jar within the jruby-complete jar and does not get loaded into a classloader until JRuby itself loads the Psych library.

I'm concerned about this going forward. Just because SnakeYAML ships a feature that can be used to load remote code does not mean any project peripherally related to SnakeYAML is exploitable, but here we are getting reports about it. Long term, one solution might be to split SnakeYAML into separate libraries for the parser and the dump/load features. The latter library is the "exploitable" code, and we do not use it in JRuby.

> I think that's the price for jruby-complete

Indeed, which is why in JRuby 9.3 we introduced an even less-bundled artifact: jruby-base, which contains only JRuby code (jruby-core existed prior to this, and bundles a few key libraries needed for basic functionality).

We can't expect everyone to use jruby-base, though, since it's not usable without a dozen other jar files also being present.

**Comment by apurtell** [comment]
> We can't expect everyone to use jruby-base, though, since it's not usable without a dozen other jar files also being present.

The problem would just move from jruby-complete to jruby-stdlib.

**Comment by asomov** [comment]
@apurtell can you please report a bug for the above mentioned "bill of material" ?
The low quality tools produce tons of false positives, they waste tons of time of the community.
They must be aware of the very negative effect of their low quality.

**Comment by asomov** [comment]
@headius for your information - it will be a major release, most probably backwards incompatible. Where can I have a look to check the compatibility and create a PR for you to simplify the migration ?

**Comment by asomov** [comment]
@headius an alternative may be to switch to [SnakeYAML Engine](https://bitbucket.org/snakeyaml/snakeyaml-engine/src/master/)
The low level API should be identical in both

**Comment by headius** [comment]
@asomov The Psych yaml library for Ruby, and the Java code we use to wrap your parser, are here: https://github.com/ruby/psych/tree/master/ext/java/org/jruby/ext/psych

The code is largely just ported from the C version, which was not terribly difficult since your low level API is very similar to libyaml. If this API remains mostly the same in SnakeYAML Engine then it shouldn't be too difficult to switch.

As always, thank you for your help and input!

**Comment by asomov** [comment]
@headius I am sorry, it is far too complex to me to contribute - I do not get how to build and test. (the documentation did not help).
I see a challenge to migrate to either new release of SnakeYAML or Engine (Java 8 will be required for both):
```
    ext.source_version = '1.7'
    ext.target_version = '1.7'
```

**Comment by apurtell** [comment]
Is it worth considering unbundling snakeyaml, bouncycastle, and other frequent fliers from jruby-stdlib? Make them dependencies of it rather than embedding them? Then we can manage jruby as easily as any other thing where a dependency must be upgraded in place. It is the bundling that requires _something_ from the jruby packager when a vulnerable dependency is embedded. Would make sense to keep jruby-complete as, well, complete; but perhaps for jruby-stdlib there are options.

**Comment by headius** [comment]
@apurtell The problem is that they are not dependencies coming in from Maven, they come in from RubyGems. Maven does not know anything about RubyGems other than via our plugins that allow fetching gems during a Maven build.

The complete and stdlib bundles just pre-install the gems, which then includes the required jar files directly inside our jar.

It might be possible to modify the way these gems locate and load the jars, such that they could look in a local repository, but it would be fragile. These jars are not simply added to the classpath; they must be loaded into our internal classloader and the library code they contain must be initialized with a running JRuby instance. They are not typical Java dependencies.

**Comment by headius** [comment]
@asomov The build and test cycle requires running JRuby, and modifying that 1.7 to be 1.8 (I am working on some cleanup on a branch now.

If you can point me toward the equivalent of ParserImpl in SnakeYAML Engine, I can take a quick look at it.

**Comment by asomov** [comment]
@headius it is the same in Engine (ParserImpl). But it implements `Iterator` interface and method `getEvent()` renamed to `next()`
Feel free to try ! The low level API is almost identical. (If this channel is inconvenient or slow, we can connect via a chat).
By the way, why not to migrate to Java 11 ? (I have only a couple of major users still under Java 8)

**Comment by headius** [comment]
A first stab at moving to SnakeYAML Engine: https://github.com/ruby/psych/pull/612

**Comment by headius** [comment]
The conversion to SnakeYAML Engine now passes all the same tests as the SnakeYAML version! Once that PR is merged and a new Psych release is out, we can upgrade and dodge the CVE concerns.

**Comment by apurtell** [comment]
@headius Understood that the embedded dependencies are not typical Maven dependencies. If they were, it would be easy to manage them. From the perspective of scan tools, they don't care. They will explode jars and find vulnerable embedded dependencies by pattern matching. I mean tools like the OWASP suite, which is very commonly utilized, especially by open source projects, and would not be characterized as "poor quality". At my workplace we use tools provided by Sonatype which function similarly. Ideally, the embedding is not done. Instead the dependencies are managed by Maven. I mean Java side dependencies like snakeyaml. For ruby dependencies the point here is the bundling and there are no issues there, other than tools which explode the jar and match on contents might find included versions of rake or rdoc or whatever that have CVE issues reported against them, and someone might come here to ask you to update and repackage those, which should be expected.

Although you may load snakeyaml in a special way for psych -- or at least, did -- in particular it is a frequent flyer these days and advised to pull out and make it a standard dependency so people won't ping you for updating an embedded version after snakeyaml drops another fix for another issue. 

> Once that PR is merged and a new Psych release is out, we can upgrade and dodge the CVE concerns.

This will be great, thank you.

**Comment by headius** [comment]
@apurtell Yes, I understand the situation, but there's no way forward with snakeyaml as a transitive dependency of the gem. Maven does not know anything about gems, nor does it know to download a jar for a gem and just put the jar file itself into the classloader as a resource. There's no way to make Maven pull SnakeYAML in such a way that the complete or stdlib jars would be able to find it.

The only option would be to have the gem logic that loads the jar file also look in the local Maven respository, but that would not work for vendored builds or nonstandard locations or build environments that don't use Maven directly.

**Comment by asomov** [comment]
@apurtell this is very sad the low quality tooling disturbs you. 
It is especially bad, because they make people make wrong conclusion like this "_after snakeyaml drops another fix for another issue._" The quasi-fix is implemented only to shut up the low quality tooling. Feel free to read the issue to have a better picture.
When CVE properly says that man with a gun is dangerous, the key warning here is the context - the gun. The low quality tooling in unable to detect the context and complain for any man. It creates enormous amount of false positives for projects like JRuby (which is not affected at all)

**Comment by apurtell** [comment]
@asomov  The tools that identify that jruby-stdlib and jruby-complete bundle snakeyaml, and that such version of snakeyaml has one or more valid CVE database entries, are in fact correct. So I don't understand how you can arrive at the conclusion it is "low quality". Perhaps you do not want to be bothered with regulatory or compliance requirements that many professionals face, and that is fine, but those of us that have such constraints will need a solution or else their jruby based component -- in our case, the HBase shell -- will be barred from production where regulatory or compliance requirements demand a clean software bill of materials (some CVSS score thresholding is typical).

**Comment by apurtell** [comment]
@headius Understood, thank you.

**Comment by asomov** [comment]
@apurtell please separate CVE and low quality tooling. CVE is correct, but it clearly defines **the context** of a possible vulnerability which is **ignored** by the low quality tooling. The low quality tooling relies only on the version. 
Please be aware that the master branch of SnakeYAML contains already the quasi-fix. It is there solely to shut up the low quality tooling. You will see that after this release of SnakeYAML the low quality tooling will remove the red flag even thought the CVE will stay ! The possibility to create custom Java instances is the feature of YAML.

**Comment by kares** [comment]
as has been said before, when a library is not used in a certain way and is contained than it should not matter, although I do understand a theoretical possibility - building a tool that just report a dependency present seems futile and wasteful.

anyhow just wanted to point out that for some gems, jars are "manageable" with [jar-dependencies](https://github.com/mkristian/jar-dependencies) and that should fulfill your "manage transitive dependency" desires. it's a bit abandoned but still works: 

- if you to have a `Jars.lock` file (in CWD) with a directive about `org.yaml:snakeyaml:...`'s version (e.g. you have the .jar on the class-path and set it to `:provided`) 
- than by the time `require 'yaml'` is executed it's `require_jar('org.yaml', 'snakeyaml', Psych::DEFAULT_SNAKEYAML_VERSION)` will effectively be a noop (the .jar packed with the gem gets ignored).

**Comment by kares** [comment]
> You can choose to adopt a friendly posture toward such professionals or not, I am not going to comment on that.

it's not about that, there's a ton of tools and I happen to see some that just check a presence of a .jar and others go a bit deeper. at my previous gig, we've seen a lot of reports, similar to this one, and at some point ended up ignoring "most" - simply due time constraints. as often tools, from user reports, where actually useless and reported weird things that were actually fixed when we tried them, while others were somehow smarter (not sure why - guessing maybe due manual intervention). 

that is the reason I suggested looking into the tooling if things can be flagged "harmless" in the tool itself.

there's actually a real-world example here with Nokogiri (which uses Xalan than has a [critical vulnearibility](https://github.com/advisories/GHSA-9339-86wc-4qgf)) - there's no way around the CVE atm and I've seen this being manually excluded from tools (as despite the .jar present the feature is [not exploitable in Nokogiri](https://github.com/sparklemotion/nokogiri/security/advisories/GHSA-qwq9-89rg-ww72)).

was just trying to help you're use-case, since you seemed to have been pushing conversation towards "transitive maven dependencies" at some point -> which actually can be done today with jar-dependencies as commented.

**Comment by headius** [comment]
https://github.com/ruby/psych/pull/612 has landed and we just need to get a release out. Once that happens, we can update JRuby.

I am a little wary of this change, though, since it means JRuby will be running a YAML 1.2 engine while everyone else is on 1.1 or lower (or possibly 1.2 if that's installed?). Might want to have this gem in the wild for a while before we set it as default.

(and of course users can still activate an earlier version with a few extra steps)

**Comment by enebo** [comment]
This will make it into 9.4.1.0 first and given no significant issues we will backport into 9.3.11.0.

**Comment by headius** [comment]
Retargeting to 9.4.1.0 since it will be released first. We may want a second bug targeting 9.3.11 so it shows up in those release notes too (though we do have a second psych 5.1 PR there already).

**Comment by floor114** [comment]
Hello @headius thank you for the fix in the 9.4.*. I see that you mentioned that 9.3.11 should include this fix as well. Are there any plans to release 9.3.11 with this fix soon?
